### PR TITLE
Separate CPU detection from SSE4 code

### LINKED ISSUE #2329: CPU detection should be separated (TU) from SSE42 code
man g++ ( -msse4.2 )

These switches enable the use of instructions in the MMX, SSE, SSE2, SSE3, SSSE3, SSE4.1, AVX, AVX2, AVX512F, AVX512PF, AVX512ER, AVX512CD, SHA, AES, PCLMUL, FSGSBASE, RDRND, F16C, FMA, SSE4A, FMA4, XOP, LWP, ABM, AVX512VL, AVX512BW, AVX512DQ, AVX512IFMA       AVX512VBMI, BMI, BMI2, FXSR, XSAVE, XSAVEOPT, LZCNT, RTM, MPX, MWAITX, PKU or 3DNow!
 extended instruction sets.  Each has a corresponding -mno- option to disable use of these instructions.         These extensions are also available as built-in functions: see x86 Built-in Functions, for details of the functions enabled and disabled by these switches. To generate SSE/SSE2 instructions automatically from floating-point code (as opposed to 387 instructions), see -mfpmath=sse. GCC depresses SSEx instructions when -mavx is used. Instead, it generates new AVX instructions or AVX equivalence for all SSEx         instructions when needed. These options enable GCC to use these extended instructions in generated code, even without -mfpmath=sse.  Applications that perform run-time CPU detection must compile separate files for each supported architecture, using the appropriate flags.  **In particular, the file containing the CPU detection code should be compiled without these options.**

```
jan@cluster9:~/arangodb
git: devel $=
» gcc -dM -E - < /dev/null | egrep "SSE|AVX" | sort
#define __SSE__ 1
#define __SSE2__ 1
#define __SSE2_MATH__ 1
#define __SSE_MATH__ 1
jan@cluster9:~/arangodb
git: devel $=
» gcc -msse4 -dM -E - < /dev/null | egrep "SSE|AVX" | sort
#define __SSE__ 1
#define __SSE2__ 1
#define __SSE2_MATH__ 1
#define __SSE3__ 1
#define __SSE4_1__ 1
#define __SSE4_2__ 1
#define __SSE_MATH__ 1
#define __SSSE3__ 1
```

https://github.com/facebook/rocksdb/blob/11c5d4741a1e11a1315d5ca644ce555e07e91f61/util/crc32c.cc#L376
https://github.com/facebook/rocksdb/blob/11c5d4741a1e11a1315d5ca644ce555e07e91f61/util/crc32c.cc#L394
https://github.com/facebook/rocksdb/blob/11c5d4741a1e11a1315d5ca644ce555e07e91f61/util/crc32c.cc#L394

### PR DESCRIPTION
**Pull request by ObiWahn**
fix #2329

### CODE REVIEW

**Comment by ObiWahn** [comment]
There were some errors submitting your CLA:

    GitHub Username is already associated with an existing contributor.
    Email is already associated with an existing contributor.

If you would like to update an existing contributor account, please email cla@fb.com with your details.

**Comment by sagar0** [comment]
@ObiWahn Can you take a look at the windows/appveyor build errors.

File: util/cpu_detect.cc [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+// four bytes at a time.
+
+#include <cstdint>
+
+namespace rocksdb {
+namespace crc32c {
+
+// Detect if SS42 or not.
+bool isSSE42() {
+#if defined(__GNUC__) && defined(__x86_64__) && !defined(IOS_CROSS_COMPILE)
```
Comment (grooverdan) [inline util/cpu_detect.cc]: why not for IOS_CROSS_COMPILE?

**Comment by grooverdan** [comment]
windows may be missing the header ```<intrin.h>``` looking at https://msdn.microsoft.com/en-us/library/hskdteyh.aspx

**Comment by ObiWahn** [comment]
changes to make / cmake will follow

**Comment by ObiWahn** [comment]
please let me know what needs to be done!

**Comment by gfosco** [comment]
Closing this via automation due to lack of activity.  If discussion is still needed here, please re-open or create a new/updated issue.

**Pull request closed without merging**